include/deal.II/differentiation/sd/symengine_scalar_operations.h

```diff
@@ -146,8 +146,317 @@ namespace Differentiation
+     *     = make_symbol_map(
+     *         Expression(...),
+     *         Tensor<1,dim,Expression>(...),
+     *         SymmetricTensor<2,dim,Expression>(...));
+     * @endcode
+     *
+     * @tparam SymbolicType Any symbolic type that is understood by the
+     *         add_to_symbol_map() functions. This includes individual
+     *         Expression, std::vector<Expression>, as well as
+     *         Tensors and SymmetricTensors of Expressions.
+     */
+    template <typename SymbolicType, typename... Args>
+    types::substitution_map
+    make_symbol_map(const SymbolicType &symbol, const Args &... other_symbols);
+
+    /**
+     * A convenience function for an adding empty entry, with the key value
+     * given by @p symbol, to the symbolic map @p substitution_map.
+     *
+     * This function is guaranteed to create an ordering that is identical
+     * to the typical add_to_substitution_map() call that is used when
+     * constructing a map to to perform symbol substitution.
+     * It exists primarily to create an initial map that can be used in the
+     * optimize() call to a BatchOptimizer, specifically if the values that
+     * are to be substituted into the map are not known at the time that the
+     * symbols used to construct symbolic expressions are defined.
+     * This helps one conform to the requirement that the arguments sent into
+     * lambda and LLVM JIT compiled functions (created by optimizing symbolic
+     * expressions) (i) be the same, and (ii) have a constant ordering.
+     *
+     * @tparam ignore_invalid_symbols A template parameter that enforces whether
+     *         or not the @p symbol has to be a valid one or not. In the
+     *         overwhelming majority of cases, the default value of
+     *         <tt>false</tt> should be selected, with the result that an
+     *         exception will be thrown if the input @p symbolic is, in fact,
+     *         not a symbolic value or expression.
+     *         An exceptional case is, for example, when performing symbolic
+     *         assembly on a finite element level. When extracting the symbolic
+     *         equivalent of the shape function gradients using FEExtractors,
+     *         the returned tensor will have some <em>a priori</em> determined
+     *         zero-valued components. These trivial components are not valid
+     *         symbols (as they are not symbolic expressions), and we would
+     *         typically wish to guard against their (erroneous) inclusion. In
+     *         this scenario, for convenience, one could set
+     *         @p ignore_invalid_symbols to <tt>true</tt> and these zero-valued
+     *         entries would be skipped over and ignored.
+     *
+     * @note In this function, the @p ValueType is somewhat arbitrary as
+     * it is only used to create dummy values.
+     */
+    template <bool ignore_invalid_symbols = false, typename ValueType = double>
```

Comment:
why could the first of these not be a regular (defaulted) function argument?

Also, maybe document `ValueType` as well?

Reply:
> why could the first of these not be a regular (defaulted) function argument?

This relates to the answer to the question below. As documented, in the overwhelming majority of cases you want this parameter to be false. However, including it in the function argument list (with a default value, the sensible choice) again necessarily dictates that it be at the end of the argument list. This breaks the method where variadic templates are used.

Reply:
I see. Bummer.